### ISSUE TITLE
The /tmp/ray directory belongs to the first person to start Ray, which leads to permission errors.

### ISSUE THREAD

**Issue by robertnishihara** [ray-project/ray#4593]
To see the issue, do the following:

1. `rm -r /tmp/ray`
2. `python -c "import ray; ray.init()"`
3. `ls -l /tmp/`

You will see a line like

```
drwxr-xr-x  3 rkn  wheel    96B Apr  9 22:31 ray/
```

which shows that the temp directory `/tmp/ray` created by Ray cannot be written by other users. If another user on the same machine tries to do `ray.init()`, they may run into errors. On the other hand, if you do

`ls -l /tmp/ray/`

you will see a line like

```
drwxrwxrwx  4 rkn  wheel   128B Apr  9 22:31 session_2019-04-09_22-31-10_72937/
```

which shows that the session directory inside of `/tmp/ray/` is writable by everyone.

**Comment by RaghuSpaceRajan** [comment]
I also see this problem when starting ray on non-head nodes in a cluster with a shared /tmp directory.

The default for a node seems to be to have /tmp/ray as the temp directory and if I specify a temp dir for a non-head node, then ray raises an error that "When connecting to an existing cluster, temp_dir must not be provided."

**Comment by gndctrl2mjrtm** [comment]
@robertnishihara I've been continuing to see this problem a lot with classrooms where students are sharing nodes. The students are having to set temp_dir in ray.init() to avoid this issue.

**Comment by robertnishihara** [comment]
cc @suquark

**Comment by RaghuSpaceRajan** [comment]
I think the intent is to have /tmp/ray writeable by everyone and its sub-directories with user-specific permissions. Is that right?

However, that still wouldn't solve the problem of non-head nodes not being allowed to have their own temp directories. Commenting out the following from worker.py should fix that:

```
        if temp_dir is not None:
            raise Exception("When connecting to an existing cluster, "
                            "temp_dir must not be provided.")

```
 Should that be a separate bug?

**Comment by gndctrl2mjrtm** [comment]
This is tricky in that tmp has to accommodate both shared and non shared systems. Could os.stat be run to check the accessibility of tmp and if it doesn't have permission, it creates a user tmp folder for stricter systems that will block ray.init otherwise?

**Comment by robertnishihara** [comment]
@gndctrl2mjrtm, do you anticipate any issues with making `/tmp/ray` as writable by everyone (but then potentially making its subfolders specific to whichever user created them)?

If that doesn't work, something like what you're suggesting could work.

@RaghuSpaceRajan that potentially looks like a different bug.

**Comment by gndctrl2mjrtm** [comment]
@robertnishihara That could work from my understanding of how Ray initializes /tmp directories.
 
However, the issue is that this has to be done across every node of the system, which makes me realize that in the situation where a user goes to a compute node and exits that node, all the tmp data is lost since it's stored on the compute node and deleted when the user leaves. 

Also, this might solve it for this particular system, but doesn't mean that other systems will not also have this whenever they have shared nodes unless they are also aware of this. 

Why can't /ray be stored in the user's ~/tmp/ray instead of /tmp/ray? That would solve this additional problem as well mentioned above.

**Comment by robertnishihara** [comment]
@gndctrl2mjrtm one setting where `/tmp/ray` makes more sense is the multi-tenant setting where a single Ray cluster is shared between many users. In that case, it's more natural for users to look in `/tmp` instead of in the home directory of whichever other user started the Ray cluster.

**Comment by gndctrl2mjrtm** [comment]
@robertnishihara What about the case such as the one we have where it's a very large number of users and shared data violates data privacy laws, so there can't be a single Ray cluster? Does the object store have the ability to distinguish whether, for instance, a user has permission to get the value back from a certain object id? Regardless of the technical feasibility of getting another correct object id, the legality might prevent me from making such a global Ray cluster. 

Not to mention having a global /tmp/ray folder where all log files are then stored, but I'm assuming that the correct permissions are automatically set so that only that user can read the logs of their own workers?

**Comment by robertnishihara** [comment]
This issue should be fixed by https://github.com/ray-project/ray/pull/4605 (once it is merged).

@gndctrl2mjrtm Currently the object store doesn't check any kind of permissions or do any authentication. Any process running on the same machine can connect to it. To make work really well in the multi-tenant setting with no sharing of information between drivers will take a lot more work. On the other hand, in this case, it might be easier to just start one Ray cluster per user.

**Comment by robertnishihara** [comment]
Fixed by #4605.
